core-libs/setup/ssr/optimized-engine/optimized-ssr-engine.ts

```diff
@@ -0,0 +1,167 @@
+import { Request, Response } from 'express';
+
+/**
+ * The rendered pages are kept in memory to be served on next request. If the `cache` is set to `false`, the
+ * response is evicted as soon as the first successful response is successfully returned.
+ */
+export class OptimizedSsrEngine {
+  protected currentConcurrency = 0;
+  protected renderingCache = new RenderingCache(this.ssrOptions);
+  private templateCache = new Map<string, string>();
+
+  get engineInstance(): NgExpressEngineInstance {
+    return this.renderResponse.bind(this);
+  }
+
+  constructor(
+    protected expressEngine: NgExpressEngineInstance,
+    protected ssrOptions?: SsrOptimizationOptions
+  ) {}
+
+  /**
+   * When SSR page can not be returned in time, we're returning index.html of
+   * the CSR application.
+   * The CSR application is returned with the "Cache-Control: no-store" response-header. This notifies external cache systems to not use the CSR application for the subsequent request.
+   */
+  protected fallbackToCsr(
+    response: Response,
+    filePath: string,
+    callback: (err?: Error | null, html?: string) => void
+  ) {
+    response.set('Cache-Control', 'no-store');
+    callback(undefined, this.getDocument(filePath));
+  }
+
+  protected getRenderingKey(request: Request): string {
+    return this.ssrOptions.renderKeyResolver
+      ? this.ssrOptions.renderKeyResolver(request)
+      : request.originalUrl;
+  }
+
+  protected getRenderingStrategy(request: Request): RenderingStrategy {
+    return this.ssrOptions.renderingStrategyResolver
+      ? this.ssrOptions.renderingStrategyResolver(request)
+      : RenderingStrategy.DEFAULT;
+  }
+
+  protected shouldRender(request: Request): boolean {
+    const concurrencyLimitExceed = this.ssrOptions.concurrency
+      ? this.currentConcurrency > this.ssrOptions.concurrency
+      : false;
+
+    return (
+      (!this.renderingCache.isRendering(this.getRenderingKey(request)) &&
+        !concurrencyLimitExceed &&
+        this.getRenderingStrategy(request) !== RenderingStrategy.ALWAYS_CSR) ||
+      this.getRenderingStrategy(request) === RenderingStrategy.ALWAYS_SSR
+    );
+  }
+
+  protected shouldTimeout(request: Request): boolean {
+    return (
+      !!this.ssrOptions?.timeout ||
+      this.getRenderingStrategy(request) === RenderingStrategy.ALWAYS_SSR
+    );
+  }
+
+  protected getTimeout(request: Request): number {
+    return this.getRenderingStrategy(request) === RenderingStrategy.ALWAYS_SSR
+      ? this.ssrOptions.forcedSsrTimeout ?? 60000
+      : this.ssrOptions.timeout;
+  }
+
+  protected returnCachedRender(
+    request: Request,
+    callback: (err?: Error | null, html?: string) => void
+  ): boolean {
+    const key = this.getRenderingKey(request);
+
+    if (this.renderingCache.isReady(key)) {
+      const cached = this.renderingCache.get(key);
+      callback(cached.err, cached.html);
+
+      if (!this.ssrOptions?.cache) {
```

Comment:
It looks to me strange that this method manipulates cache. Looks like a just in case solution, where you don't trust the render mechanism as it might sometimes save to cache even when it should not.

Reply:
Should be more obvious with updated documentation for the cache option.